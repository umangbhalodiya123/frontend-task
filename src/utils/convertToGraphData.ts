import { Employee } from '@/types'

export const convertToGraphData = (data: Employee[], key: keyof Employee) => {
  const graphData: { label: string; value: number }[] = []

  const labels = [...new Set(data.map((employee) => employee[key]))].toSorted(
    (a, b) =>
      typeof a === 'number' && typeof b === 'number'
        ? a - b
        : typeof a === 'string' && typeof b === 'string'
        ? a.localeCompare(b)
        : 0
  )

  labels.forEach((label) => {
    const avgPerformanceRating =
      data
        .filter((employee) => employee[key] === label)
        .reduce((acc, curr) => acc + curr.performance_rating, 0) /
      data.filter((employee) => employee[key] === label).length

    graphData.push({
      label: String(label),
      value: avgPerformanceRating,
    })
  })

  return graphData
}
