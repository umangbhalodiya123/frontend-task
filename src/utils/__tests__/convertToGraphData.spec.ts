import { convertToGraphData } from '../convertToGraphData'
import { employees } from '../'

describe('convertToGraphData', () => {
  it('converts data correctly for department', () => {
    const convertedData = convertToGraphData(employees, 'department')
    expect(convertedData).toMatchSnapshot()
  })

  it('converts data correctly for team_size', () => {
    const convertedData = convertToGraphData(employees, 'team_size')
    expect(convertedData).toMatchSnapshot()
  })

  it('converts data correctly for age', () => {
    const convertedData = convertToGraphData(employees, 'age')
    expect(convertedData).toMatchSnapshot()
  })
})
