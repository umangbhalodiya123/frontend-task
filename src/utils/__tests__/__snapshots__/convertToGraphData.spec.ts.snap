// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertToGraphData converts data correctly for age 1`] = `
[
  {
    "label": "21",
    "value": 2,
  },
  {
    "label": "23",
    "value": 6.666666666666667,
  },
  {
    "label": "24",
    "value": 4.5,
  },
  {
    "label": "26",
    "value": 8,
  },
  {
    "label": "31",
    "value": 4,
  },
  {
    "label": "37",
    "value": 5,
  },
  {
    "label": "39",
    "value": 9,
  },
]
`;

exports[`convertToGraphData converts data correctly for department 1`] = `
[
  {
    "label": "Business Development",
    "value": 7,
  },
  {
    "label": "Legal",
    "value": 4,
  },
  {
    "label": "Product Management",
    "value": 2,
  },
  {
    "label": "Research and Development",
    "value": 8,
  },
  {
    "label": "Sales",
    "value": 5,
  },
  {
    "label": "Services",
    "value": 5.666666666666667,
  },
  {
    "label": "Training",
    "value": 7,
  },
]
`;

exports[`convertToGraphData converts data correctly for team_size 1`] = `
[
  {
    "label": "3",
    "value": 7,
  },
  {
    "label": "6",
    "value": 2,
  },
  {
    "label": "7",
    "value": 8,
  },
  {
    "label": "9",
    "value": 4.666666666666667,
  },
  {
    "label": "11",
    "value": 5,
  },
  {
    "label": "15",
    "value": 9,
  },
  {
    "label": "17",
    "value": 5,
  },
]
`;
